src/main/java/com/eternitywall/ots/OpenTimestamps.java

```diff
@@ -410,46 +414,52 @@ public static boolean upgrade(Timestamp timestamp) {
         Set<TimeAttestation> existingAttestations = timestamp.getAttestations();
         for (Timestamp subStamp : timestamp.directlyVerified()) {
             for (TimeAttestation attestation : subStamp.attestations) {
-                if (attestation instanceof PendingAttestation) {
+                if (attestation instanceof PendingAttestation && !subStamp.isTimestampComplete()) {
+
                     String calendarUrl = new String(((PendingAttestation) attestation).getUri(), StandardCharsets.UTF_8);
                     // var calendarUrl = calendarUrls[0];
                     byte[] commitment = subStamp.msg;
 
-                    Calendar calendar = new Calendar(calendarUrl);
-                    Timestamp upgradedStamp = OpenTimestamps.upgrade(subStamp, calendar, commitment, existingAttestations);
-                    if(upgradedStamp != null) {
+                    try {
+                        Calendar calendar = new Calendar(calendarUrl);
+                        Timestamp upgradedStamp = OpenTimestamps.upgrade(subStamp, calendar, commitment, existingAttestations);
                         try {
                             subStamp.merge(upgradedStamp);
                         } catch (Exception e) {
                             e.printStackTrace();
                         }
                         upgraded = true;
-                        return upgraded;
+                    }catch(Exception e){
+                        log.info(e.getMessage());
                     }
                 }
             }
         }
         return upgraded;
     }
 
-    private static Timestamp upgrade(Timestamp subStamp, Calendar calendar, byte[] commitment, Set<TimeAttestation> existingAttestations) {
-        Timestamp upgradedStamp = calendar.getTimestamp(commitment);
-        if (upgradedStamp == null) {
-            return null;
+    private static Timestamp upgrade(Timestamp subStamp, Calendar calendar, byte[] commitment, Set<TimeAttestation> existingAttestations) throws Exception{
+
+        Timestamp upgradedStamp;
+        try {
+            upgradedStamp = calendar.getTimestamp(commitment);
+            if (upgradedStamp == null) {
+                throw new Exception("Invalid stamp");
+            }
+        }catch(Exception e){
```

Comment:
Why not specific Exception?
ExceptionA | ExceptionB notation should be supported